cohortextractor/generate_codelist_report.py

```diff
@@ -0,0 +1,109 @@
+import os
+
+import pandas as pd
+import structlog
+
+from .codelistlib import codelist_from_csv
+from .mssql_utils import mssql_dbapi_connection_from_url
+from .tpp_backend import make_batches_of_insert_statements, standardise_if_date
+
+logger = structlog.get_logger()
+
+
+def generate_codelist_report(output_dir, codelist_path, start_date, end_date):
+    """
+    Generate a pair of CSV files reporting usage of codes in given codelist.
+
+        * counts_per_code counts how many times each code appears in the given timeframe
+        * counts_per_week counts how many times all the codes appear in each week in the
+            given timeframe
+
+    Weeks begin on a Monday, and a code appearing on (say) a Wednesday is counted as
+    belonging to the following Monday.
+    """
+
+    codelist = codelist_from_csv(codelist_path, "snomedct")
+    start_date = standardise_if_date(start_date)
+    end_date = standardise_if_date(end_date)
+
+    queries = codelist_queries(codelist) + [
+        events_query(start_date, end_date),
+        population_query(start_date, end_date),
+        results_query(),
+    ]
+
+    connection = mssql_dbapi_connection_from_url(os.environ["DATABASE_URL"])
+    cursor = connection.cursor()
+
+    for query in queries:
+        logger.debug(query)
+        cursor.execute(query)
+
+    counts = pd.DataFrame(list(cursor), columns=["code", "date", "num"])
+
+    counts_per_code = counts.groupby("code")["num"].sum()
+    counts_per_week = counts.groupby(pd.Grouper(key="date", freq="W-MON"))["num"].sum()
+
+    os.makedirs(output_dir, exist_ok=True)
+
+    counts_per_code.to_csv(output_dir / "counts_per_code.csv")
+    counts_per_week.to_csv(output_dir / "counts_per_week.csv")
+
+
+def codelist_queries(codelist):
+    # These are SNOMED-specific values
+    collation = "Latin1_General_BIN"
+    max_code_len = 18
+
+    queries = [
+        f"""
+        CREATE TABLE #codelist (
+          code VARCHAR({max_code_len}) COLLATE {collation},
+          category VARCHAR(MAX)
+        )
+        """
+    ]
+    rows = [[code, ""] for code in codelist]
+    queries += make_batches_of_insert_statements(
+        "#codelist", ("code", "category"), rows
+    )
+    return queries
+
+
+def events_query(start_date, end_date):
+    return f"""
+    SELECT * INTO #events FROM (
+        SELECT Patient_ID, ConceptID, ConsultationDate
+        FROM CodedEvent_SNOMED
+        INNER JOIN #codelist
+            ON CodedEvent_SNOMED.ConceptID = #codelist.code
+        WHERE ConsultationDate BETWEEN '{start_date}' AND '{end_date}'
+    ) t
+    """
+
+
+def population_query(start_date, end_date):
+    # We're interested in patients who are registered on end_date, or who have died
+    # between start_date and end_date.
```

Comment:
Given that we're counting coded events rather than patients it's not obvious to me why we need to worry much about the population at all. But I can well believe there's a good reason for it. Just wanted to check it was deliberate decision and hadn't been cargo-culted from elsewhere!

Reply:
This is deliberate, but thanks for checking!

Reply:
There's been quite a bit of discussion about the population used. So, at the request of the researchers, we're including only those patients mentioned. It might well change, but for now we're keeping it the same as the query we're comparing it with.

It might be relevant to the occurrence of codes because apparently, when someone leaves (or presumably joins) a practice there's an increase in codes, as they make sure their routine checks are done and any repeat prescriptions issued, etc, before they go. I suspect it probably evens out though!

Reply:
Right yes, that makes sense. Thanks.